libclc/clc/lib/generic/math/clc_hypot.inc

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2014 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+// Returns sqrt(x*x + y*y) with no overflow or underflow unless the result
+// warrants it
+
+#if __CLC_FPSIZE == 32
+_CLC_DEF _CLC_OVERLOAD __CLC_GENTYPE __clc_hypot(__CLC_GENTYPE x,
+                                                 __CLC_GENTYPE y) {
+  __CLC_UINTN ux = __CLC_AS_UINTN(x);
+  __CLC_UINTN aux = ux & EXSIGNBIT_SP32;
+  __CLC_UINTN uy = __CLC_AS_UINTN(y);
+  __CLC_UINTN auy = uy & EXSIGNBIT_SP32;
+  __CLC_INTN c = aux > auy;
+  ux = c ? aux : auy;
+  uy = c ? auy : aux;
+
+  __CLC_INTN xexp = __clc_clamp(
+      __CLC_AS_INTN(ux >> EXPSHIFTBITS_SP32) - EXPBIAS_SP32, -126, 126);
+  __CLC_GENTYPE fx_exp =
+      __CLC_AS_GENTYPE((xexp + EXPBIAS_SP32) << EXPSHIFTBITS_SP32);
+  __CLC_GENTYPE fi_exp =
+      __CLC_AS_GENTYPE((-xexp + EXPBIAS_SP32) << EXPSHIFTBITS_SP32);
```

Comment:
This should use frexp's exponent to extract the exponent instead of bithacking (or the utility discussed on one of the previous PRs)

Reply:
That would be good. I'll need to think about how best to go about it as there's quite a few examples of this pattern scattered around libclc, with various addends etc., like `xexp`  and `-xexp` here.